src/Documentation/resources/Best_Practices.md

```diff
@@ -0,0 +1,421 @@
+
+-   The initial State.ReadStateAsync() automatically occurs prior to
+    ActiveAsync() has been called for a grain.
+
+-   When the grain’s state object’s data is changed, then the grain should call
+    State.WriteStateAsync()
+
+    -   Typically, grains call State.WriteStateAsync() at the end of grain
+        method to return the Write promise.
+
+    -   The Storage provider *could* try to batch Writes that may increase
+        efficiency, but behavioral contract and configurations are orthogonal
+        (independent) to the storage API used by the grain.
+
+    -   A **timer** is an alternative method to write updates periodically.
+
+        -   The timer allows the application to determine the amount of
+            “eventual consistency”/statelessness allowed.
+
+        -   Timing (immediate/none/minutes) can also be controlled as to when to
+            update.
+
+    -   Each grain class can only be associated with one storage provider.
+
+        -   [StorageProvider(ProviderName=”name”)] attribute associates the
+            grain class with a particular provider
+
+        -   \<StorageProvider\> will need to be added to the Silo config file
+            which should also include the corresponding “name” from
+            [StorageProvider(ProviderName=”name”)]
+
+        -   A composite storage provider can be used with SharedStorageProvider
+
+## Storage Providers
+
+Built-in Storage Providers
+
+-   Orleans.Storage houses all of the built-in storage providers. The namespace
+    is: OrleansProviders.dll
+
+-   MemoryStorage (Data stored in memory without durable persistence) is used
+    *only* for debugging and unit testing.
+
+-   AzureTableStorage
+
+    -   Configure the Azure storage account information with an optional
+        DeleteStateOnClear (hard or soft deletions)
+
+    -   Orleans serializer efficiently stores binary data in one Azure table
+        cell
+
+    -   Data size limit == max size of the Azure column which is 64kb of binary
+        data
+
+    -   Community contributed code that extends the use of multiple table
+        columns which increases the overall maximum size to 1mb.
+
+-   SharedStorageProvider will write data to multiple underlying storage
+    providers by using the grain id hash.
+
+    -   Example: *need to create*
+
+Storage Provider Debugging Tips
+
+-   TraceOverride Verbose3 will log much more information about storage
+    operations.
+
+    -   Update silo config file
+
+        -   LogPrefix=”Storage” for all providers, or specific type using
+            “Storage.Memory” / ”Storage.Azure” / “Storage.Shard”
+
+How to deal with Storage Operation Failures
+
+-   Grains and storage providers can await storage operations and *retry*
+    failures as needed
+
+-   Unhandled failures will propagate back to the caller and will be seen by the
+    client as a broken promise
+
+-   Other than the initial read, there is not a concept that automatically
```

Comment:
If grain lets an `InconsistentStateException` escape, it gets automatically deactivated.